core/src/commonMain/kotlin/com/saveourtool/processbuilder/ProcessBuilder.kt

```diff
@@ -44,104 +47,88 @@ class ProcessBuilder(private val useInternalRedirections: Boolean, private val f
-            logger.debug { "stderr of `$command`:\t${stderr.joinToString("\t")}" }
+        if (executionResult.stderr.isNotEmpty()) {
+            logger.debug { "stderr of `$command`:\t${executionResult.stderr.joinToString("\t")}" }
         }
-        redirectTo?.let {
-            fs.write(redirectTo) {
-                write(stdout.joinToString("\n").encodeToByteArray())
-                write(stderr.joinToString("\n").encodeToByteArray())
-            }
-        } ?: logger.trace { "Execution output:\t$stdout" }
-        return ExecutionResult(status, stdout, stderr)
+
+        return executionResult
     }
 
     private fun modifyCmd(
         command: String,
-        directory: String,
+        directory: Path?,
         processBuilderInternal: ProcessBuilderInternal,
-    ): String {
-        // If we need to step out into some directory before execution
-        val cdCmd = if (directory.isNotBlank()) {
-            if (isCurrentOsWindows()) {
-                "cd /d $directory && "
-            } else {
-                "cd $directory && "
-            }
-        } else {
-            ""
-        }
-        // Additionally process command for Windows, it it contain `echo`
-        val commandWithEcho = cdCmd + if (isCurrentOsWindows()) {
-            processCommandWithEcho(command)
-        } else {
-            command
-        }
-        logger.trace { "Modified cmd: $commandWithEcho" }
-        // Finally, make platform dependent adaptations
-        return processBuilderInternal.prepareCmd(commandWithEcho)
-    }
+    ): String = command
+        .let(::processCommandWithEchoOnWindows)
+        .also { cmd -> logger.trace { "Processed command with echo for windows: $cmd" } }
+        .let { changeWorkingDirectory(command, directory) }
+        .also { cmd -> logger.trace { "Processed command for directory change: $cmd" } }
+        .let(processBuilderInternal::appendShellCommand)
+        .also { cmd -> logger.trace { "Redirected stdout and stderr: $cmd" } }
+        .let(processBuilderInternal::runCommandByNameOfAnotherUser)
+        .also { cmd -> logger.trace { "Modified command to run by name of user ${config.childProcessUsername}: $cmd" } }
 
-    /**
-     * Log error message and throw exception
-     *
-     * @param errMsg error message
-     * @throws ProcessExecutionException
-     */
-    private fun logErrorAndThrowProcessBuilderException(errMsg: String): Nothing {
-        logger.error { errMsg }
-        throw ProcessExecutionException(errMsg)
+    private fun processCommandWithEchoOnWindows(command: String): String = when {
+        isCurrentOsWindows() -> processCommandWithEcho(command)
+        else -> command
     }
 
+    private fun changeWorkingDirectory(command: String, directory: Path?): String = when {
+        directory == null -> ""
+        directory.segments.isEmpty() -> "".also { logger.warn { "Requested working directory " } }
+        isCurrentOsWindows() -> "cd /d $directory && "
```

Comment:
`cd` is a `cmd.exe` builtin. It will not work this way.

In both case (Windows and UNIX), you'll need a `boolean` flag to support launching a program via `exec()` and via the default shell (`${SHELL}` or `%COMSPEC%`).

 - When launching via `exec()`, the new CWD is merely passed to the library call as a parameter.
 - When launching via the shell, platform-specific tricks like `cmd.exe /C "cd dir/ && command.exe"` will be needed.